retrier.go

```diff
@@ -284,6 +283,51 @@ func (r *Retrier) DoWithContext(ctx context.Context, callback func(*Retrier) err
 	}
 }
 
+// DoFunc1 is a helper for retrying callback functions that return a value or an
+// error. It returns the last value returned by a call to callback, and reports
+// an error if none of the calls succeeded.
+// (Note this is not a method of Retrier, since methods can't be generic.)
+func DoFunc1[T any](ctx context.Context, r *Retrier, callback func(*Retrier) (T, error)) (T, error) {
```

Comment:
```suggestion
// DoFunc is a helper for retrying callback functions that return a value or an
// error. It returns the last value returned by a call to callback, and reports
// an error if none of the calls succeeded.
// (Note this is not a method of Retrier, since methods can't be generic.)
func DoFunc[T any](ctx context.Context, r *Retrier, callback func(*Retrier) (T, error)) (T, error) {
```

My suggestion is to call this just `DoFunc` so that users that want to use a struct to handle multiple return values can pretend the other ones don't exist.

Not sure how widespread it is, but I like the convention that when you're providing multiple methods instead of variadicity, the first one doesn't have a number, but the latter ones do. Cf in [Haskell](https://hackage.haskell.org/package/base-4.19.1.0/docs/Data-List.html#g:17) there's `zip`, `zip3`, `zip4`, where `zip` is implicitly a `zip2`.